### ISSUE TITLE
[Patterns] What should distributing a heap mean?

### ISSUE THREAD

**Issue by lydia-duncan** [chapel-lang/chapel#18966]
This issue is part of a series of issues to design the interface for collections across serial, parallel, and distributed contexts. Our goal is to determine what we think is reasonable to support and what doesn't seem useful. Additional patterns welcome (but it would be best to put them in their own issue for discussion, likely)

What does a distributed heap mean?  Is the whole thing the heap, meaning that for any given `pop` call the result could involve a remote read?  Is it run on a per-locale basis, so `pop` would give the maximum/minimum for that locale but not necessarily for the heap as a whole?  Is the "top" X of the heap replicated across locales so the answer is the same from any locale but the user is assumed to not call `pop` in parallel, to give the data structure time to synchronize the top again?  Should `push` calls be batched until a good locale guess for them can be determined?  Should the user be able to specify what kind of distributed heap they want to suit their use case?

**Comment by mppf** [comment]
Is there a distributed heap implementation already? Are we trying to design a distributed heap API?

I think it would depend on the use case for the distributed heap but I would probably expect each locale having a different heap top to be something more like a distributed collection containing heaps. So if we had something called a distributed heap I would expect it to have a single top across all locales. But I don't know that we need that data structure at all in the near term.

**Comment by lydia-duncan** [comment]
I did not see a distributed heap implementation in the languages I surveyed (C++, Java, Swift, Go, and Rust), but we were asked to think about parallel and distributed implementations for all our collection types.  I'm fine with "we won't support it" but I think we should at least be thinking about it now as part of this larger design effort.

**Comment by mppf** [comment]
I'd look at the UPC++ collections for prior work on a distributed data structures, FWIW

**Comment by vasslitvinov** [comment]
The paper "Cone: A Distributed Heap Approach to Resource Selection" seems to have a single max element in the heap.

In general, I expect a data structure's semantic definition to remain the same whether it is serial, parallel, or distributed. For example, if it is a distributed set, the user should not be able to observe multiple equivalent keys. Or the same key mapped to multiple values in a distributed map. Unless it is a multimap.

The interesting question is how to make the usual heap operations efficient. For example, what API we can offer that can do aggregation, like aggregating a bunch of of push or pop operations, behind the scenes.

Another interesting question is "memory consistency". What happens if one task pushes when another task concurrently pops? What if a lot of tasks distributed over a lot of locales keep pop()-ing concurrently? If the behavior of the last scenario is trivial semantically, what if these tasks synchronize with each other - does it mean that the updates to the heap must also synchronize?

**Comment by mppf** [comment]
> Another interesting question is "memory consistency".

I have been thinking for a long time that data structures should be able to do their own caching (think caching like `--cache-remote` - for example accumulating updates) and that when they do that it would be cool to allow the data structures somehow to be aware of memory fences that occur. (Here I am talking about the memory fences already defined - say with sequential-consistency atomics or sync variables or task start/end.). That way, when two tasks synchronize, the data structure can e.g. flush the accumulated updates. But, the data structure can operate more efficiently when there is a long region of no synchronization. A possible alternative strategy would be to use the same idea but make the fences part of the data structure API rather than hooking them into the existing synchronization.

**Comment by bradcray** [comment]
I agree with Michael's original comment.  I don't personally feel the need for us to work on a distributed heap in the short-term.

**Comment by lydia-duncan** [comment]
> I'd look at the UPC++ collections for prior work on a distributed data structures, FWIW

UPC++ doesn't seem to list collections in its documentation, it just has an example of how to implement a distributed map using its functionality.  Do you have a link I'm missing?  I also searched for the types individually and looked in its extension documentation

**Comment by mppf** [comment]
I was talking about this -- https://github.com/berkeley-container-library/bcl
